README.md
# MMM-GoogleAnalytics

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/) to display data from Google Analytics.

![Screenshot of a MMM-GoogleAnalytics](http://robstechlog.com/wp-content/uploads/2017/07/MMM-Googleanalytics_feature.png?raw=true)

<br>

An optional broadcast of visitors/customers location data to [MMM-Globe](https://github.com/SvenSommer/MMM-Globe) is also possible.<p>

![Screenshot of a MMM-GoogleAnalytics](https://github.com/SvenSommer/MMM-GoogleAnalytics/blob/master/MMM-GoogleAnalytics_animated.gif?raw=true)
##### Important note to MMM-Globe
Unfortunately MMM-Globe is not capable of running properly on a Raspberry Pi 3 causes of lacking computing power.

Or with the words of the creator of MMM-Globe: "Due to the intensive calculations needed to construct the objects, and the computational ability needed, I do not recommend using this module if your Mirror uses a Raspberry Pi. Please only use this module if you have something a little more powerful driving your mirror. I don’t want my work to be responsible for any mishaps with your devices."

## Using the module
### 1. Install the module

In your terminal, go to your MagicMirror's Module folder:
````
cd ~/MagicMirror/modules
````

Clone this repository:
````
git clone https://github.com/SvenSommer/MMM-GrafanaChart
````

### 2. Install Googles Analytics Core Reporting API and node-geocoder library
Install the Googles Analytics Core Reporting API for Node.js in your ``MMM-GoogleAnalytics`` folder by using npm:

```
cd MMM-GoogleAnalytics
npm install --save googleapis
npm install node-geocoder
```
### 3. Get the credentials from Google
To get the credentials from Google, following this [blogpost](http://robstechlog.com/2017/07/19/display-website-statistics-smart-mirror-mmm-googleanalytics/).

After you have done this, you should be able to:
* Copy your ``viewID``. You'll need it in the next step.
* Rename your credential file (e.g. ``mywebsiteGAapi-6116b1dg49a1.json``) in ``keyfile.json`` and save it in your ``MMM-GoogleAnalytics`` folder, like ``MMM-GoogleAnalytics/keyfile.json``. If you haven't done already.

### 4. Configure the module in your `config.js` file.
To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
        module: 'MMM-GoogleAnalytics',
        header: 'mywebsite.com - Today',
        position: 'top_left',
        config: {
            viewID : 'ga:123456700', // see README.md, how to get viewID and your keyfile.json
            start_date: 'today', //today,1daysAgo,
            end_date: 'today',
            metrics: 'ga:newusers, ga:users, ga:sessions, ga:pageviews,  ga:sessionDuration',
			dimensions: 'ga:country', //ga:country, ga:city,ga:source
			sort: '-ga:timeOnPage',
			filters: 'ga:timeOnPage>10', //ga:timeOnPage>10 Return results where the time on the page is strictly greater than ten seconds.
			segment: '',
			start_index: 1,
			max_results: 50,
			updateInterval: 1 * 10 * 1000, // every minute
			showtable: true,
            showtotalline: true,
			exportdatatoMMM_Globe: false,
			debug: false
            }
        },
    ]
}
```
### 5. (Optional) Configure the module to broadcast to MMM-Globe
If you want to display the geographic  location of your website visitors on the [MMM-Globe](https://github.com/SvenSommer/MMM-Globe) module, follow the following steps:
* The Query needs to include ``'ga:city, ga:country'`` for the dimensions. In this specific order.
* Follow the install instructions of [MMM-Globe](https://github.com/SvenSommer/MMM-Globe)
* within the ``config section`` of MMM-Globe, change ``receiveExternalLocations:`` to ``1``.

## Configuration options
Check out the [common queries](https://developers.google.com/analytics/devguides/reporting/core/v3/common-queries) in the [API Reference](https://developers.google.com/analytics/devguides/reporting/core/v3/reference).

The [Query Explorer](https://ga-dev-tools.appspot.com/query-explorer/) lets you play with the API by building queries. Very useful!

| Option           | Required | Type | Description
|----------------- |-----------|-------|----------------
| `viewID`        | *yes*|`string`| Google Analytics view (profile) ID. Follow this [blogpost](http://robstechlog.com/2017/07/19/display-website-statistics-smart-mirror-mmm-googleanalytics/) to get it.
| `start_date`        | *yes*|`string`| Start date for fetching Analytics data.
| `end_date`        | *yes*|`string`| End date for fetching Analytics data.
| `metrics`        | *yes*|`string`| The aggregated statistics for user activity to your site, such as clicks or pageviews. See [Possible Options](https://developers.google.com/analytics/devguides/reporting/core/dimsmets).
| `dimensions`        | *no*|`string`| Breaks down metrics by common criteria.
| `sort`        | *no*|`string`| A list of metrics and dimensions indicating the sorting order and sorting direction for the returned data.
| `filters`        | *no*|`string`| restricts the data returned from your request.
| `max_results`        | *no*|`int`| The maximum number of rows to include in the response.
| `showtable` | *no*|`bool`| `true` - shows results in a table. `false` - no table is shown
| `showtotalline` | *no*|`bool`| `true` - shows a sum in the end of each column. `false` - no summary on each column
| `exportdatatoMMM_Globe` | *no*|`bool`| `true` - Modul sends coordinates of users to another module called [MMM-Globe](https://github.com/SvenSommer/MMM-Globe). Therefore you need to include ``ga:city, ga:country`` in dimension section of your query.
| `debug` | *no*|`bool`| `true` - shows debug information. `false` - no debug info is shown


CHANGELOG.md
## [1.0.0] - Initial Release

First public release

## [1.1.0] - several improvements

* Added Line for totals
* Times are now formated in hh:mm:ss
* foolowing Parameters are now optional: 		 
    * dimensions
    * sort
    * filters
    * segment
    * start_index
    * max_results

## [1.2.0] - Broadcasting Geodata implemented

* Google Maps is queried for Coordinates of the City, Country
* Coorrdinates are send via Notification and could be used to display on MMM-Globe

## [1.3.0] - Broadcasting Geodata implemented

* Implemented optional symbols for Header information in table view
* Bugfix: automatic reload reworked

## [1.3.1] - additonal optional Parameter added

* showtable, showtotalline and debug parameter added
* added Debug mode
* Bugfix: automatic reload reworked


node_helper.js
/* Magic Mirror
 * Node Helper: MMM-GoogleAnalytics
 *
 * By SvenSommer
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");
var googleapi = require('googleapis');
var ApiKeyFile = require('./keyfile.json');
var NodeGeocoder = require('node-geocoder');
var fs = require('fs');
var config;
var wcities = [];
var text;

module.exports = NodeHelper.create({

	// Override socketNotificationReceived method.

	/* socketNotificationReceived(notification, payload)
	 * This method is called when a socket notification arrives.
	 *
	 * argument notification string - The identifier of the noitication.
	 * argument payload mixed - The payload of the notification.
	 */
	socketNotificationReceived: function(notification, payload) {
		var self = this;
		this.config = payload;
		if (self.config.debug) {
			 console.log(notification, "received by node_helper.js received");
		}
		if (notification === "MMM-GoogleAnalytics-QUERY_DATA") {
			this.authorize(function(data){
					self.sendNotification_DISPLAY_DATA(data);
			});
		}
		else if (notification === "MMM-GoogleAnalytics-GET_COORDINATES") {
			//console.log("MMM-GoogleAnalytics-GET_COORDINATES received");
			var self = this;
			this.cities = payload;
			this.getCoordinatesfromCitiesName(this.cities, function(cityinfo){
				//console.log(cityinfo);
				self.sendSocketNotification("MMM-GoogleAnalytics-CITYINFO", cityinfo);
			});
		}
	},

	// Example function send notification test
	sendNotification_DISPLAY_DATA: function(payload) {
		this.sendSocketNotification("MMM-GoogleAnalytics-DISPLAY_DATA", payload);
	},

	getCoordinatesfromGoogle: function(city,callback){
		var geocoder = NodeGeocoder({provider: 'google'});
		//console.log("query geocoder for ",city.country + "," + city.cityname);

			geocoder.geocode({address: city.country + "," + city.cityname, minConfidence: 0.5, limit: 5}, function(err, res) {
				if (err) {
					console.log(err);
					return;
				}
				var cityobject= {
					'country':   city.country,
					'cityname':  city.cityname,
					'latitude':  res[0].latitude,
					'longitude': res[0].longitude}

					callback(cityobject);

			});



	},

	getCoordinatesfromCitiesName: function(cities,callback){
		self = this;
		for (var x = 0, ln = cities.length; x < ln; x++) {
		  setTimeout(function(y) {
			 // console.log(cities[y].country ,cities[y].cityname);
			  if (self.cities[y].country != "(not set)" && self.cities[y].cityname != "(not set)") {
				  self.getCoordinatesfromGoogle(self.cities[y],function(cityinfo){
					  callback(cityinfo);
				  });
			  }
		  }, x * 500, x);
		}
	},

	writeJsonfile: function(cities){
		var fs = require('fs');
		var file = fs.createWriteStream('test.json');
		file.write("MMM_GoogleAnalytics_Users = [");
		for (var cityinfo in cities) {
			if (object.hasOwnProperty(city)) {
				file.write("{lat:"+ cityinfo.latitude+",lng:"+cityinfo.longitude+", label:" + cityinfo.cityname + "},");
			}
		}
		file.write("]");
		file.end();
	},

	authorize: function(callback) {
		var	self = this;
		var google = this.getdefaultObj(googleapi);
		var Key = this.getdefaultObj(ApiKeyFile);

		var jwtClient = new google.default.auth.JWT(Key.default.client_email, null, Key.default.private_key, ['https://www.googleapis.com/auth/analytics.readonly'], null);
		jwtClient.authorize(function (err, tokens) {
		  if (err) {
		    console.log(err);
		    return;
		  }
		  var analytics = google.default.analytics('v3');

		  var querybuilder =  {
			  'auth': jwtClient,
			  'ids': self.config.viewID ,
			  'metrics': self.config.metrics,
			  'start-date': self.config.start_date,
	  	  	  'end-date': self.config.end_date,
		  };
		  if (self.config.dimensions != '') {querybuilder['dimensions'] = self.config.dimensions;}
		  if (self.config.sort != '') {querybuilder['sort'] = self.config.sort;}
		  if (self.config.filters != '') {querybuilder['filters'] = self.config.filters;}
		  if (self.config.segment != '') {querybuilder['segment'] = self.config.segment;}
		  if (self.config.start_index != '') {querybuilder['start-index'] = self.config.start_index;}
		  if (self.config.max_results != '') {querybuilder['max-results'] = self.config.max_results;}

		  if (self.config.debug) {
		  	console.log("query data:",querybuilder);
		  }

		 analytics.data.ga.get(
			 querybuilder

		  , function (err, response) {
			if (err) {
			  console.log(err);
			  return;
			}
			//console.log(JSON.stringify(response, null, 4));
			callback(response);
		  });
		});
	},

	getdefaultObj: function(obj) {
		return obj && obj.__esModule ? obj : { default: obj };
	}


});


MMM-GoogleAnalytics.js
/* global Module */

/* Magic Mirror
 * Module: MMM-GoogleAnalytics
 *
 * By SvenSommer
 * MIT Licensed.
 */

Module.register("MMM-GoogleAnalytics", {
	defaults: {

		viewID : 'ga:123456700',
		metrics: 'ga:newusers,ga:users, ga:sessions, ga:pageviews',
		start_date: 'today',
		end_date: 'today',
		dimensions: 'ga:country',
		sort: '-ga:pageviews',
		max_results: 10,
		filters: '',
		animationSpeed: 5000,
		updateInterval: 10 * 10 * 1000, //every ten minutes
		retryDelay: 500000,
		showtotalline: true,
		showtable: true,
		displaySymbols: true,
		exportdatatoMMM_Globe: false,
		debug: false
	},

	requiresVersion: "2.1.0", // Required version of MagicMirror

	start: function() {
		var self = this;
		var dataRequest = null;
		var dataNotification = null;

		//Flag for check if module is loaded
		this.loaded = false;
		this.getData();

		//this.sendTestArray();

		setInterval(function() {
			self.getData()
			self.updateDom();
		}, this.config.updateInterval);
	},

	processData: function() {
		var self = this;
		//this.dataRequest = data;
		if (this.loaded === false) { self.updateDom(self.config.animationSpeed) ; }
		this.loaded = true;


	},

	/*
	 * getData
	 * function example return data and show it in the module wrapper
	 * get a URL request
	 *
	 */
	getData: function() {
		var self = this;
		//this.loaded = false;
		this.sendSocketNotification("MMM-GoogleAnalytics-QUERY_DATA", this.config);
		this.loaded = true;	



	},
	/* scheduleUpdate()
	 * Schedule next update.
	 *
	 * argument delay number - Milliseconds before next update.
	 *  If empty, this.config.updateInterval is used.
	 */
	scheduleUpdate: function(delay) {
		var nextLoad = this.config.updateInterval;
		if (typeof delay !== "undefined" && delay >= 0) {
			nextLoad = delay;
		}
		nextLoad = nextLoad ;
		var self = this;
		setTimeout(function() {
			self.getData();

		}, nextLoad);
	},

	getDom: function() {
		var self = this;

		// create element wrapper for show into the module
		var wrapper = document.createElement("div");
		wrapper.className = "small";
		if(!this.loaded) {
                wrapper.innerHTML = "Loading...";
				wrapper.classname = "small dimmed";
                return wrapper;
        }

		if (this.dataNotification) {
			if (this.debug) {
				console.log(this.dataNotification);
			}


			if (this.config.showtable) {
				var wrapperDataNotification = document.createElement("div");

					var table = document.createElement("table");
					table.className = "small";

					var hrow = document.createElement("tr");
					var trow = document.createElement("tr");
					var columnDataTypes = [];


					for (var h in this.dataNotification.columnHeaders) {
						var cheader_element = this.dataNotification.columnHeaders[h];
						columnDataTypes.push(cheader_element.dataType);
						//header line
						var headercell = document.createElement("td");
						if (!this.config.displaySymbols) {
							headercell.innerHTML = this.translate(cheader_element.name);
							headercell.className = "header";
						}else {
							var headeraymbol = document.createElement("span");
							if (!this.translate(cheader_element.name).indexOf("fa-")) {
								headercell.innerHTML = cheader_element.name;
							}else {
								headeraymbol.className = this.translate(cheader_element.name);
							}
							headercell.appendChild(headeraymbol);
							headercell.className =  " centered";
							}
						hrow.appendChild(headercell);


						//total line
						var totalcell = document.createElement("td");

						if (cheader_element.dataType === "INTEGER" ){
							totalcell.innerHTML  = this.dataNotification.totalsForAllResults[cheader_element.name];
							totalcell.className = "totalline centered";
						}
						else if (cheader_element.dataType === "TIME") {
							totalcell.innerHTML  = this.getTimeFormat(this.dataNotification.totalsForAllResults[cheader_element.name]);
							totalcell.className = "totalline align-right";
						}
						else {
							totalcell.innerHTML = "";
							totalcell.className = "totalline";
						}
						trow.appendChild(totalcell);

					}
					table.appendChild(hrow);


					//rows
					for (var r in this.dataNotification.rows) {
						var crow_element  = this.dataNotification.rows[r];
						var row = document.createElement("tr");
						i = 0;
						for (var c in crow_element) {
							var rowcell = document.createElement("td");
							if (columnDataTypes[i] === "TIME") {
								rowcell.innerHTML = this.getTimeFormat(crow_element[c]);
								rowcell.className = "align-right";
							}else {
								rowcell.innerHTML = crow_element[c];
								rowcell.className = "centered";
							}
							row.appendChild(rowcell);
							i++;
						}
						table.appendChild(row);
					}
					//adding total line
					if (this.config.showtotalline) {
						table.appendChild(trow);
					}

					wrapper.appendChild(table);
				}

				if (this.config.debug) {
					var d = new Date();
					var labelLastUpdate = document.createElement("label");
					labelLastUpdate.innerHTML = "Updated: " + ("0" + d.getHours()).slice(-2) + ":" + ("0" + d.getMinutes()).slice(-2)+ ":" + ("0" + d.getSeconds()).slice(-2) + "<br>Intervall: " + this.config.updateInterval/1000 + "s";
					wrapper.appendChild(labelLastUpdate);
				}


			}
		return wrapper;
	},

	exportDatatoOtherModule: function(){
		var cities =  [];
		for (var r in this.dataNotification.rows) {
				var crow_element  = this.dataNotification.rows[r];
				cities.push({
				'cityname': crow_element[0],
				'country':crow_element[1]
				});
			}
			this.getCoordinatesfromCitiesName(cities);
	},

	getTimeFormat: function(int) {
		var sec_num = parseInt(int, 10); // don't forget the second param
	    var hours   = Math.floor(sec_num / 3600);
	    var minutes = Math.floor((sec_num - (hours * 3600)) / 60);
	    var seconds = sec_num - (hours * 3600) - (minutes * 60);

	    if (hours   < 10 && hours > 0) {hours   = "0"+hours + "h :";} else if (hours === 0) {hours  = ""} else {hours   = hours + "h :";};
	    if (minutes < 10) {minutes = "0"+minutes;}
	    if (seconds < 10) {seconds = "0"+seconds;}

	    return hours+minutes+'m '+seconds+'s ';
	},

	getScripts: function() {
		return [];
	},

	getCoordinatesfromCitiesName: function(cities){
		this.sendSocketNotification("MMM-GoogleAnalytics-GET_COORDINATES", cities);

	},

	getStyles: function () {
		return [
			"MMM-GoogleAnalytics.css", "font-awesome.css"
		];
	},

	// Load translations files
	getTranslations: function() {
		if (!this.config.displaySymbols) {
			return {
				en: "translations/en.json",
				es: "translations/es.json",
				de: "translations/de.json",

			};
		}else {
			return {
				en: "translations/symbols.json",
				es: "translations/symbols.json",
				de: "translations/symbols.json",
			};
		}


	},

	// socketNotificationReceived from helper
	socketNotificationReceived: function (notification, payload) {
		if (this.debug) {
			console.log(notification, "received by MMM-GoogleAnalytics.js received");
		}

		if(notification === "MMM-GoogleAnalytics-DISPLAY_DATA") {
			this.dataNotification = payload;
			this.updateDom();
			this.loaded = true;
			if ( this.config.exportdatatoMMM_Globe) {
				this.exportDatatoOtherModule();
			}
		}
		else if (notification === "MMM-GoogleAnalytics-CITYINFO") {
			this.sendNotification("MMM-GoogleAnalytics-CITYINFO", payload);
		}
	},
	notificationReceived: function (notification, cityinfo, sender) {
		if(notification === "MMM-GoogleAnalytics-CITYINFO" && sender === "MMM-GoogleAnalytics") {
			if (this.debug) {
				console.log("MMM-GoogleAnalytics-CITYINFO received by MMM-GoogleAnalytics", cityinfo);
			}
		}
	},
});
